protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -246,32 +246,24 @@ impl NetworkBehaviour for Behaviour {
         local_addr: &Multiaddr,
         remote_addr: &Multiaddr,
     ) -> Result<THandler<Self>, ConnectionDenied> {
-        match self
-            .outgoing_direct_connection_attempts
-            .remove(&(connection_id, peer))
+        if is_relayed(local_addr) {
+            return Ok(Either::Left(handler::relayed::Handler::new(
+                ConnectedPoint::Listener {
+                    local_addr: local_addr.clone(),
+                    send_back_addr: remote_addr.clone(),
+                },
+            ))); // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
+        }
+
+        if let Some(&relayed_connection_id) = self.direct_to_relayed_connections.get(&connection_id)
```

Comment:
I don't think this can ever be hit actually. In hole-punching, we cannot for sure tell which inbound connection was the one that the remote achieved through hole punching.

Plus, all this state tracking here uses `ConnectionId`s that are created as part of dialing, i.e. outbound connections.

So unless I am missing something, this function can be reduced to always return a `dummy::ConnectionHandler` for non-relayed connections.

@mxinden Thoughts?

Reply:
I think I agree with your assessment. I'll make this change after @mxinden's comments.

Reply:
You are right @thomaseizinger.

For what it is worth, likely my train of thought that ended up with the code above:

In `NetworkBehaviour::on_connection_handler_event` on `OutboundConnectNegotiated` we trigger a `Dial` with `role_override()` (i.e. `as_listener`) and then insert the `maybe_direct_connection_id` from the `DialOpts` into `self.direct_to_relayed_connections`.

https://github.com/libp2p/rust-libp2p/blob/d4c4078e6dad6c4ac1f18a084c35fa6e91016d84/protocols/dcutr/src/behaviour_impl.rs#L379-L395

On success this would be reported to the `Behaviour` via `NetworkBehaviour::handle_established_outbound_connection` and not as assumed here in `NetworkBehaviour::handle_established_inbound_connection`.

Reply:
How about adding an `assert` here, that `connection_id` is not in `self.direct_to_relayed_connections`?